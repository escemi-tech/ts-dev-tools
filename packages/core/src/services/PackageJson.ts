import { PackageJson as PackageJsonType } from "type-fest";

import { PackageJsonMerge } from "./PackageJsonMerge";
import { FileService } from "./FileService";
import { join } from "path";

export type JsonArray = boolean[] | number[] | string[] | JsonFileData[] | Date[];
export type AnyJson = boolean | number | string | JsonFileData | Date | JsonArray | JsonArray[];

export type JsonFileData = {
  [key: string]: AnyJson | undefined;
};
export type PackageJsonContent = JsonFileData & PackageJsonType;

export class PackageJson {
  private static readonly PACKAGE_JSON_FILE_NAME = "package.json";

  private content?: PackageJsonContent = undefined;

  constructor(private path: string) {
    if (!FileService.fileExists(this.path)) {
      throw new Error(`Package.json "${this.path}" does not exist`);
    }
  }

  getPath(): string {
    return this.path;
  }

  getContent(): PackageJsonContent {
    if (this.content) {
      return this.content;
    }
    return (this.content = JSON.parse(FileService.getFileContent(this.path)) as PackageJsonContent);
  }

  setContent(content: PackageJsonContent): void {
    this.content = content;
    this.write();
  }

  getPackageName(): string | undefined {
    return this.getContent().name;
  }

  getPackageVersion(): string | undefined {
    return this.getContent().version;
  }

  isPrivate(): boolean {
    return this.getContent().private === true;
  }

  getTsDevToolsVersion(): string | undefined {
    const tsDevToolsConfig = this.getContent().tsDevTools as JsonFileData | undefined;
    const version = tsDevToolsConfig?.version as string | undefined;
    return version;
  }

  getDependenciesPackageNames(): string[] {
    const dependencies = this.getContent().dependencies;
    return dependencies ? Object.keys(dependencies) : [];
  }

  getDevDependenciesPackageNames(): string[] {
    const devDependencies = this.getContent().devDependencies;
    return devDependencies ? Object.keys(devDependencies) : [];
  }

  getAllDependenciesPackageNames(): string[] {
    return Array.from(
      new Set([...this.getDependenciesPackageNames(), ...this.getDevDependenciesPackageNames()])
    );
  }

  merge(update: PackageJsonContent): void {
    this.content = PackageJsonMerge.merge(this.getContent(), update);
    this.write();
  }

  backup(): string {
    const backupPath = this.path + ".backup";
    FileService.copyFile(this.path, backupPath);
    return backupPath;
  }

  restore(backupPath: string): void {
    FileService.copyFile(backupPath, this.path);
    this.content = undefined;
  }

  private write() {
    FileService.putFileContent(this.path, JSON.stringify(this.content, null, 2));
  }

  static fromDirPath(dirPath: string): PackageJson {
    const packageJsonPath = join(dirPath, PackageJson.PACKAGE_JSON_FILE_NAME);
    return new PackageJson(packageJsonPath);
  }
}
