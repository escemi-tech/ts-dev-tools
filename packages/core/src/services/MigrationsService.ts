import { readdirSync, unlinkSync } from "fs";
import { resolve } from "path";

import { PackageJson } from "../services/PackageJson";
import { Plugin, PluginService } from "./PluginService";

export type Migration = { fullname: string; shortname: string; path: string };

export type MigrationUpFunction = (absoluteProjectDir: string) => Promise<void>;

export class MigrationsService {
  static MIGRATION_BUILT_PATH = "dist/install/migrations";

  static async executeMigrations(
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Promise<void> {
    const migrations = MigrationsService.getAvailableMigrations(absoluteProjectDir, currentVersion);

    const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const packageJsonBackupPath = packageJson.backup();

    try {
      for (const migration of migrations) {
        console.info(`Applying migration "${migration.fullname}"...`);

        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { up } = require(migration.path);

        // Apply migration
        await up(absoluteProjectDir);

        // Upgrade current version
        PackageJson.fromDirPath(absoluteProjectDir).merge({
          tsDevTools: { version: migration.shortname },
        });

        console.info(`Migration "${migration.fullname}" applied!`);
      }
    } catch (error) {
      // Rollback package.json
      packageJson.restore(packageJsonBackupPath);
      throw error;
    }

    unlinkSync(packageJsonBackupPath);
  }

  private static getAvailableMigrations(
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Migration[] {
    const installedPlugins = PluginService.getInstalledPlugins(absoluteProjectDir);

    const migrationFiles: Migration[] = [];
    for (const installedPlugin of installedPlugins) {
      migrationFiles.push(
        ...MigrationsService.getPluginMigrations(installedPlugin, currentVersion)
      );
    }

    migrationFiles.sort(({ shortname: nameA }, { shortname: nameB }) => nameA.localeCompare(nameB));

    return Array.from(new Set(migrationFiles));
  }

  private static getPluginMigrations(
    plugin: Plugin,
    currentVersion: string | undefined
  ): Migration[] {
    const migrationFiles = [];

    const pluginMigrationsDirPath = resolve(plugin.path, MigrationsService.MIGRATION_BUILT_PATH);
    for (const migrationFile of readdirSync(pluginMigrationsDirPath)) {
      if (!migrationFile.match(/^[0-9]{14}-[-a-z]+\.(js|ts)$/)) {
        continue;
      }

      const migrationName = MigrationsService.getMigrationNameFromFile(migrationFile);
      const shouldApplyMigration = MigrationsService.migrationIsAfterCurrentVersion(
        migrationName,
        currentVersion
      );

      if (!shouldApplyMigration) {
        continue;
      }

      migrationFiles.push({
        shortname: migrationName,
        fullname: `${plugin.shortname} - ${migrationName}`,
        path: resolve(pluginMigrationsDirPath, migrationFile),
      });
    }
    return migrationFiles;
  }

  public static getMigrationNameFromFile(migrationFile: string): string {
    const migrationName = migrationFile.split(".").slice(0, -1).join(".");
    return migrationName;
  }

  public static migrationIsAfterCurrentVersion(migrationName: string, currentVersion?: string) {
    return !currentVersion || currentVersion.localeCompare(migrationName) < 0;
  }
}
