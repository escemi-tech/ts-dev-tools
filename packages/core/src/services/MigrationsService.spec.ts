import { getConsoleInfoContent, mockConsoleInfo, resetMockedConsoleInfo } from "../tests/console";
import { createProjectForTestFile, deleteTestProject } from "../tests/test-project";
import { MigrationsService } from "./MigrationsService";
import { PackageJson } from "./PackageJson";

// Set to false to avoid using the cache
const useCache = true;
// Set to false to inspect the test project directory after the test
const shouldCleanupAfterTest = true;

describe("MigrationsService", () => {
  let testProjectDir: string;

  beforeAll(() => {
    if (!useCache) {
      console.warn("Cache is disabled. Enable it one dev is done.");
    }
    if (!shouldCleanupAfterTest) {
      console.warn("Cleanup is disabled. Enable it one dev is done.");
    }
  });

  beforeEach(async () => {
    testProjectDir = await createProjectForTestFile(__filename, useCache);
    mockConsoleInfo();
  });

  afterEach(async () => {
    resetMockedConsoleInfo();
    if (shouldCleanupAfterTest) {
      await deleteTestProject(__filename);
    }
  });

  afterAll(() => {});

  describe("executeMigrations", () => {
    it("should execute migrations when no version is provided", async () => {
      PackageJson.fromDirPath(testProjectDir).merge({
        devDependencies: {
          "@ts-dev-tools/core": "1.0.0",
        },
      });

      const executeMigrationsAction = () =>
        MigrationsService.executeMigrations(testProjectDir, undefined);

      await expect(executeMigrationsAction()).resolves.toBeUndefined();

      expect(getConsoleInfoContent()).toMatchSnapshot();
    });

    it("should execute migrations when no available migrations exist after given version", async () => {
      PackageJson.fromDirPath(testProjectDir).merge({
        devDependencies: {
          "@ts-dev-tools/core": "1.0.0",
        },
      });

      const executeMigrationsAction = () =>
        MigrationsService.executeMigrations(testProjectDir, "20201024173398-init");

      await expect(executeMigrationsAction()).resolves.toBeUndefined();
      expect(getConsoleInfoContent()).toMatchSnapshot();
    });
  });

  describe("getMigrationNameFromFile", () => {
    it("should retrieve migration name from given file name", async () => {
      const migrationFile = "20201024173398-test.ts";
      const migrationName = MigrationsService.getMigrationNameFromFile(migrationFile);

      expect(migrationName).toEqual("20201024173398-test");
    });
  });

  describe("migrationIsAfterCurrentVersion", () => {
    it("should retrieve true when migration name is after current version", async () => {
      const migrationName = "20210130173398-other-test";
      const currentVersion = "20201024173398-test";

      const migrationIsAfterCurrentVersion = MigrationsService.migrationIsAfterCurrentVersion(
        migrationName,
        currentVersion
      );
      expect(migrationIsAfterCurrentVersion).toBe(true);
    });

    it("should retrieve false when migration name is before current version", async () => {
      const migrationName = "20201024173398-test";
      const currentVersion = "20210130173398-other-test";

      const migrationIsAfterCurrentVersion = MigrationsService.migrationIsAfterCurrentVersion(
        migrationName,
        currentVersion
      );
      expect(migrationIsAfterCurrentVersion).toBe(false);
    });

    it("should retrieve true when current version is undefined", async () => {
      const migrationName = "20201024173398-test";

      const migrationIsAfterCurrentVersion =
        MigrationsService.migrationIsAfterCurrentVersion(migrationName);
      expect(migrationIsAfterCurrentVersion).toBe(true);
    });
  });
});
