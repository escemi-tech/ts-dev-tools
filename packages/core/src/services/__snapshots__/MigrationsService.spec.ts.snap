// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MigrationsService executeMigrations should execute migrations when no available migrations exist after given version 1`] = `
"Applying migration \\"core - 20220617100200-prettier-cache\\"...
Migration \\"core - 20220617100200-prettier-cache\\" applied!"
`;

exports[`MigrationsService executeMigrations should execute migrations when no version is provided 1`] = `
"Applying migration \\"core - 20201024173398-init\\"...
Migration \\"core - 20201024173398-init\\" applied!
Applying migration \\"core - 20220617100200-prettier-cache\\"...
Migration \\"core - 20220617100200-prettier-cache\\" applied!"
`;
